File: LeetCode-2023/面试题 17.05. 字母与数字.cpp
class Solution {
public:
    vector<string> findLongestSubarray(vector<string>& array) {
        int n = array.size();
        int s[n + 1];
        s[0] = 0;
        for(int i = 0; i < n; i++) {
            s[i + 1] = s[i] + (array[i][0] >> 6 & 1) * 2 - 1;
        }
        int left = 0, right =0;
        unordered_map<int, int> first;
        for(int i = 0; i < n + 1; i++) {
            auto it = first.find(s[i]);
            if(it == first.end()) {
                first[s[i]] = i;
            } else {
                if(i - it->second > right - left) {
                    left = it->second;
                    right = i;
                }
            }
        }
        return {array.begin() + left, array.begin() + right};
    }
};

File: LeetCode-2023/面试题 17.05. 字母与数字.py
class Solution:
    def findLongestSubarray(self, array: List[str]) -> List[str]:
        prefixSums = list(accumulate((-1 if v[0].isdigit() else 1 for v in array), initial=0))
        left = right = 0
        first = {}
        for index, prefixSum in enumerate(prefixSums):
            it = first.get(prefixSum, -1)
            if it == -1:
                first[prefixSum] = index
            else:
                if index - it > right - left:
                    right = index
                    left = it
        return array[left:right]